packages/protocol/lib/compatibility/ast-version.ts

```diff
@@ -0,0 +1,76 @@
+// tslint:disable: max-classes-per-file
+import { ContractVersion, ContractVersionDelta, ContractVersionDeltaIndex, ContractVersions } from '@celo/protocol/lib/compatibility/version'
+import {
+  BuildArtifacts,
+} from '@openzeppelin/upgrades'
+
+/**
+ * A version report for a specific contract.
+ */
+export class ASTContractVersionReport {
+  constructor(public readonly oldVersion: ContractVersion, public readonly newVersion: ContractVersion, public readonly expectedDelta: ContractVersionDelta) {}
+
+  public expectedVersion = (): ContractVersion => {
+    return this.expectedDelta.appliedTo(this.oldVersion)
+  }
+
+  public matches = (): boolean => {
+    return this.newVersion.toString() === this.expectedVersion().toString()
+  }
+}
+
+/**
+ * A mapping {contract name => {@link ASTVersionReport}}.
+ */
+export interface ASTContractVersionReportIndex {
+  [contract: string]: ASTContractVersionReport
+}
+
+export class ASTContractVersionsReport {
+  static create = async (oldArtifacts: BuildArtifacts, newArtifacts: BuildArtifacts, expectedVersionDeltas: ContractVersionDeltaIndex): Promise<ASTContractVersionsReport> => {
+    const oldVersions = await ContractVersions.fromArtifacts(oldArtifacts)
+    const newVersions = await ContractVersions.fromArtifacts(newArtifacts)
+    const contracts = {}
+    Object.keys(newVersions.contracts).map((contract:string) => {
+      const versionDelta = expectedVersionDeltas[contract] === undefined ? ContractVersionDelta.fromChanges(false, false, false, false) : expectedVersionDeltas[contract]
+      const oldVersion = oldVersions.contracts[contract] === undefined ? ContractVersion.fromString('0.0.0.0') : oldVersions.contracts[contract]
```

Comment:
Does this imply that when adding a new contract that has no storage, its first version number would be 0.0.1.0 ? (Nothing wrong about that, just checking)

Reply:
Yeah great question, I think this should actually be `1.0.0.0`, as the storage/code compatibility report should come back clean, which implies no version delta, which implies the actual version should be the same as what we have here.